cupy/cuda/memory.pyx

```diff
@@ -800,41 +808,41 @@ cdef class SingleDeviceMemoryPool:
                         else:
                             total = size + self.total_bytes()
                             raise OutOfMemoryError(size, total)
-            chunk = Chunk(mem, 0, size, stream_ptr)
+            chunk = _Chunk.__new__(_Chunk)
+            chunk._init(mem, 0, size, stream_ptr)
 
         assert chunk.stream_ptr == stream_ptr
         rlock.lock_fastrlock(self._in_use_lock, -1, True)
         try:
             self._in_use[chunk.ptr] = chunk
         finally:
             rlock.unlock_fastrlock(self._in_use_lock)
-        if remaining is not None:
-            self._append_to_free_list(remaining.size, remaining, stream_ptr)
         pmem = PooledMemory(chunk, self._weakref)
         return MemoryPointer(pmem, 0)
 
     cpdef free(self, size_t ptr, Py_ssize_t size):
         cdef set free_list
-        cdef Chunk chunk
+        cdef _Chunk chunk
 
         rlock.lock_fastrlock(self._in_use_lock, -1, True)
         try:
-            chunk = self._in_use.pop(ptr, None)
+            chunk = self._in_use.pop(ptr)
+        except KeyError:
+            raise RuntimeError('Cannot free out-of-pool memory')
         finally:
             rlock.unlock_fastrlock(self._in_use_lock)
-        if chunk is None:
-            raise RuntimeError('Cannot free out-of-pool memory')
         stream_ptr = chunk.stream_ptr
 
         if chunk.next is not None:
             if self._remove_from_free_list(chunk.next.size, chunk.next,
                                            stream_ptr):
-                chunk = self._merge(chunk, chunk.next)
+                chunk.merge(chunk.next)
 
         if chunk.prev is not None:
             if self._remove_from_free_list(chunk.prev.size, chunk.prev,
                                            stream_ptr):
-                chunk = self._merge(chunk.prev, chunk)
+                chunk = chunk.prev
+                chunk.merge(chunk.next)
```

Comment:
looks this merge `chunk.prev` with `chunk.next`? we need to merge `chunk.prev` with `chunk`.

Reply:
No, this merge `original_chunk.prev` with `original_chunk.prev.next` that is `original`.

Reply:
Ah, okay